scylla-macros/src/from_row.rs

```diff
@@ -14,19 +14,18 @@ pub(crate) fn from_row_derive(tokens_input: TokenStream) -> Result<TokenStream,
     // Generates a token that sets the values of struct fields: field_type::from_cql(...)
     let (fill_struct_code, fields_count) = match struct_fields {
         crate::parser::StructFields::Named(fields) => {
-            let set_fields_code = fields.named.iter().map(|field| {
+            let set_fields_code = fields.named.iter().enumerate().map(|(col_ix, field)| {
                 let field_name = &field.ident;
                 let field_type = &field.ty;
 
                 quote_spanned! {field.span() =>
                     #field_name: {
-                        let (col_ix, col_value) = vals_iter
-                            .next()
-                            .expect("BUG: Size validated iterator did not contain the expected number of values"); 
+                        // to avoid unnecessary copy `std::mem::take` is used
+                        let col_value = ::std::mem::take(&mut row_columns[#col_ix]);
```

Comment:
`expect` (a method that may panic) is replaced with indexing (which may panic as well). This defeats the purpose of `expect_used` lint, which is to keep the code free of possible panics.

Reply:
I've used an array instead of iterator `let mut row_columns: [_; #fields_count]`, which allows for Rust compiler to detect such `index out of range` errors during the compile time.
So technically speaking if you have not disabled unconditional panic lint (which you definitely not), this code is safe.
And if in this code a bug will be appears it will be `99.9%` (`0.1%` stand when unconditional panic will be disabled, and I dont why someone really want to do that, its like to disable a borrow checker) catches by the compiler, with the detailed explanation what was incorrect.

E.g. I've modified this line of the code to
```Rust
let col_value = ::std::mem::take(&mut row_columns[#col_ix + 1]);
```
And rustc immediately complains
```
error: this operation will panic at runtime
    --> scylla-cql/src/frame/response/cql_to_rust.rs:1014:43
     |
1014 |         struct MyRow(i32, Option<String>, Option<Vec<i32>>);
     |                                           ^^^^^^ index out of bounds: the length is 3 but the index is 3
     |
     = note: `#[deny(unconditional_panic)]` on by default

error: could not compile `scylla-cql` (lib test) due to 1 previous error
```

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_lint/builtin/static.UNCONDITIONAL_PANIC.html

Reply:
So to sum up, this code allows to catch an error on the compile time, using `.expect` does not allow to do that and will return it during the runtime only

Reply:
You're right, it seems I have missed that you made `row_columns` an array. In such case indexing can be indeed checked by the compiler, with no panic risk in runtime if no warning is emitted about that.
Could you please write in comments why the array is used (to preserve length info in the type, to avoid panics) and why indexing won't panic (because array has a compile-time known length, and the index is a compile-time const, too)?

Reply:
@wprzytula done